### PR TITLE
Supports for DETR position embedding

### PR DESCRIPTION
**Pull request by v-qjqs**
This PR implements DETR position embedding.

### CODE REVIEW

File: mmdet/models/transformer_head/position_encoding.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+import math
+
+import torch
+import torch.nn as nn
+from mmcv.cnn import uniform_init
+
+
+class PositionEmbeddingSine(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/transformer_head/position_encoding.py:8]: Move this file to mmdet/models/utils/

File: mmdet/models/transformer_head/position_encoding.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    """Position encoding with sine and cosine functions.
+
+    See 'End-to-End Object Detection with Transformers'
```
Comment (ZwwWayne) [inline mmdet/models/transformer_head/position_encoding.py:11]:
```python
`End-to-End Object Detection with Transformers
    <https://arxiv.org/pdf/2005.12872>`_

```

File: mmdet/models/transformer_head/position_encoding.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                 scale=None,
+                 eps=1e-6):
+        super(PositionEmbeddingSine, self).__init__()
+        self.num_pos_feats = num_pos_feats
+        self.temperature = temperature
+        self.normalize = normalize
+        if scale is not None and not normalize:
+            raise ValueError('normalize should be True if scale is passed')
+        if scale is None:
+            scale = 2 * math.pi
```
Comment (ZwwWayne) [inline mmdet/models/transformer_head/position_encoding.py:41]: Why don't we directly set `scale=2*math.pi` as default?

File: mmdet/models/transformer_head/position_encoding.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        scale (float): A scale factor for calculating the position
+            embedding. Default None. If None, 2*pi will be used.
+        eps (float): A value added to the denominator for numerical
+            stability. Default 1e-6.
+    """
+
+    def __init__(self,
+                 num_pos_feats,
+                 temperature=10000,
+                 normalize=False,
```
Comment (ZwwWayne) [inline mmdet/models/transformer_head/position_encoding.py:31]: argument `normalize` is unnecessary because it is determined by `scale`.

File: mmdet/models/utils/position_encoding.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        """Forward function for `PositionEmbeddingSine`.
+
+        Args:
+            mask (Tensor): ByteTensor mask. Non-zero values representing
+                ignored positions, while zero values means valid positions
+                for this image.
+
+        Returns:
+            pos (Tensor): Returned position embedding with shape
+                [bs,num_pos_feats*2,h,w].
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:55]: Change the format according to previous comments.

File: mmdet/models/utils/position_encoding.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
+import math
+
+import torch
+import torch.nn as nn
+from mmcv.cnn import uniform_init
+
+
+class PositionEmbeddingSine(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:8]: PositionEmbeddingSine -> SinePositionEmbedding

File: mmdet/models/utils/position_encoding.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        repr_str = self.__class__.__name__
+        repr_str += f'(num_pos_feats={self.num_pos_feats}, '
+        repr_str += f'temperature={self.temperature}, '
+        repr_str += f'normalize={self.normalize}, '
+        repr_str += f'scale={self.scale}, '
+        repr_str += f'eps={self.eps})'
+        return repr_str
+
+
+class PositionEmbeddingLearned(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:88]: PositionEmbeddingLearned -> LearnedPositionEmbedding

File: mmdet/models/utils/position_encoding.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(self,
+                 num_pos_feats,
+                 temperature=10000,
+                 normalize=False,
+                 scale=2 * math.pi,
+                 eps=1e-6):
+        super(SinePositionEmbedding, self).__init__()
+        if normalize:
+            assert isinstance(scale, float), 'when normalize is set, scale '\
+                f'should be provided and type float , found {type(scale)}'
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:38]: and in float type

File: mmdet/models/utils/position_encoding.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+    def __init__(self,
+                 num_pos_feats,
+                 temperature=10000,
+                 normalize=False,
+                 scale=2 * math.pi,
+                 eps=1e-6):
+        super(SinePositionEmbedding, self).__init__()
+        if normalize:
+            assert isinstance(scale, float), 'when normalize is set, scale '\
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:37]: float -> (float,  int)

File: mmdet/models/utils/position_encoding.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    See `End-to-End Object Detection with Transformers
+    <https://arxiv.org/pdf/2005.12872>`_ for details.
+
+    Args:
+        num_pos_feats (int): The feature dimension for each position
+            along x-axis or y-axis. Note the final returned dimension
+            for each position is 2 times of this value.
+        temperature (int): The temperature used for calculating the
+            position embedding. Default 10000.
+        normalize (bool): Whether to do normalization when calculating
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:23]: Whether to normalize the position embedding.

File: mmdet/models/utils/position_encoding.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    """Position encoding with sine and cosine functions.
+
+    See `End-to-End Object Detection with Transformers
+    <https://arxiv.org/pdf/2005.12872>`_ for details.
+
+    Args:
+        num_pos_feats (int): The feature dimension for each position
+            along x-axis or y-axis. Note the final returned dimension
+            for each position is 2 times of this value.
+        temperature (int): The temperature used for calculating the
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:21]: The temperature used for scaling the position embedding.

File: mmdet/models/utils/position_encoding.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            position embedding. Default 10000.
+        normalize (bool): Whether to do normalization when calculating
+            the position embedding. Default False.
+        scale (float): A scale factor for calculating the position
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:25]: A scale factor that scales the position embedding. The `scale` will be used only when `normalize=True`. Default 2*pi.

File: mmdet/models/utils/position_encoding.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+@POSITION_ENCODING.register_module()
+class LearnedPositionEmbedding(nn.Module):
+    """Position embedding with learnable embedding weights.
+
+    Args:
+        num_pos_feats (int): The feature dimension for each position
+            along x-axis or y-axis. The final returned dimension for
+            each position is 2 times of this value.
+        row_num_embed (int): The dictionary size of row embeddings.
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:100]: Add default

File: mmdet/models/utils/position_encoding.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        col_num_embed (int): The dictionary size of col embeddings.
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:101]: 1. col_num_embed (int, optional)
2. Add default.

File: mmdet/models/utils/position_encoding.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        super(LearnedPositionEmbedding, self).__init__()
+        self.row_embed = nn.Embedding(row_num_embed, num_pos_feats)
+        self.col_embed = nn.Embedding(col_num_embed, num_pos_feats)
+        self.num_pos_feats = num_pos_feats
+        self.row_num_embed = row_num_embed
+        self.col_num_embed = col_num_embed
+        self.init_weights()
+
+    def init_weights(self):
+        uniform_init(self.row_embed)
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:114]: Docstring

File: tests/test_models/test_position_encoding.py [hunk 16]
```diff
@@ -0,0 +1,37 @@
+import pytest
+import torch
+
+from mmdet.models.utils import LearnedPositionEmbedding, SinePositionEmbedding
+
+
+def test_position_encoding_sine(num_pos_feats=16, batch_size=2):
+    with pytest.raises(AssertionError):
```
Comment (ZwwWayne) [inline tests/test_models/test_position_encoding.py:8]: What assertion is tested?

File: mmdet/models/utils/position_encoding.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+                 temperature=10000,
+                 normalize=False,
+                 scale=2 * math.pi,
+                 eps=1e-6):
+        super(SinePositionEmbedding, self).__init__()
+        if normalize:
+            assert isinstance(scale, (float, int)), 'when normalize is set,' \
+                'scale should be provided and in float or int type, ' \
+                f'found {type(scale)}'
+        self.num_pos_feats = num_pos_feats
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:43]: Can we simply use `num_feats` rather than `num_pos_feats`?

File: tests/test_models/test_position_encoding.py (hunk 16 above)
Comment (v-qjqs) [reply tests/test_models/test_position_encoding.py:8]: The type of scale which is not in [int,float] will raise assertion error.

Comment (ZwwWayne) [reply tests/test_models/test_position_encoding.py:8]: Please add this as a comment in the code.

File: mmdet/models/utils/position_encoding.py [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+    See `End-to-End Object Detection with Transformers
+    <https://arxiv.org/pdf/2005.12872>`_ for details.
+
+    Args:
+        num_feats (int): The feature dimension for each position
+            along x-axis or y-axis. Note the final returned dimension
+            for each position is 2 times of this value.
+        temperature (int): The temperature used for scaling the position
+            embedding. Default 10000.
+        normalize (bool): Whether to normalize the position embedding.
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:23]: If the argument has a default value, optional should be added in the type, like `(bool, optional)`, `(float, optional)`. Please check other places in the code and modify it accordingly.

File: mmdet/models/utils/position_encoding.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+    Args:
+        num_feats (int): The feature dimension for each position
+            along x-axis or y-axis. Note the final returned dimension
+            for each position is 2 times of this value.
+        temperature (int, optional): The temperature used for scaling
+            the position embedding. Default 10000.
+        normalize (bool, optional): Whether to normalize the position
+            embedding. Default False.
+        scale (float, optional): A scale factor that scales the position
+            embedding. The scale will be used only when normalize is True.
```
Comment (ZwwWayne) [inline mmdet/models/utils/position_encoding.py:26]:
```python
normalize -> `normalize`
```

**Review by ZwwWayne — APPROVED** [review]
